substrafl_examples/go_further/plot_substrafl_sklearn_fedavg.py

```diff
@@ -0,0 +1,543 @@
+            # If shared state is None, we are at the init state of the algorithm.
+            # We need all different instances of the algorithm to have the same
+            # initialization.
+            self._model.coef_ = np.ones((OUTPUT_SIZE, INPUT_SIZE))
+            self._model.intercept_ = np.zeros(3)
+        else:
+            # If we have a shared state, we update the model parameters with
+            # the average parameters updates.
+            self._model.coef_ += np.reshape(
+                shared_state.avg_parameters_update[:-1],
+                (OUTPUT_SIZE, INPUT_SIZE),
+            )
+            self._model.intercept_ += shared_state.avg_parameters_update[-1]
+
+        # To be able to compute the delta between the parameters before and after training,
+        # we need to save them in a temporary variable.
+        old_coef = self._model.coef_
+        old_intercept = self._model.intercept_
+
+        # Model training.
+        self._model.fit(datasamples["data"], datasamples["targets"])
+
+        # We compute de delta.
+        delta_coef = self._model.coef_ - old_coef
+        delta_bias = self._model.intercept_ - old_intercept
+
+        # We reset the model parameters to their state before training in order to remove
+        # the local updates from it.
+        self._model.coef_ = old_coef
+        self._model.intercept_ = old_intercept
+
+        # We output the len of the dataset to apply a pondered average between
+        # the organizations regarding their number of samples, and the local
+        # parameters updates.
+        # These updates are sent to the aggregator to compute the average
+        # parameters updates, that we will receive in the next round in the
+        # `shared_state`.
+        return fl_schemas.FedAvgSharedState(
+            n_samples=len(datasamples["targets"]),
+            parameters_update=[p for p in delta_coef] + [delta_bias],
+        )
+
+    @remote.remote_data
+    def predict(self, datasamples, shared_state, predictions_path):
+        """The predict function to be executed on organizations containing
+        data we want to test our model on. The @remote_data decorator is mandatory
+        to allow this function to be sent and executed on the right organization.
+
+        Args:
+            datasamples (_type_): datasamples extracted from the organizations data using
+                the given opener.
+            shared_state (_type_): shared_state provided by the aggregator.
+            predictions_path (_type_, optional): Path where to save the predictions.
+                This path is provided by Substra and the metric will automatically
+                get access to this path to load the predictions.
+        """
+        predictions = self._model.predict(datasamples["data"])
+
+        if predictions_path is not None:
+            np.save(predictions_path, predictions)
+            shutil.move(str(predictions_path) + ".npy", predictions_path)
```

Comment:
Why do we have to remove the ".npy" here? I remember having to the same thing in the opener back in the days.

Reply:
Cause if you load an npy with numpy, it makes some assumption that will make the loading fails... It's really weird but you're right, this is present in several places in substrafl

Reply:
Done :)